Journalist can create article and login using devise

Joberg95 opened the pull request:
#### PT Story: https://www.pivotaltracker.com/story/show/164490825
#### Description
Added user model with devise
Added role journalist
Login routes to index after log in
Added create article page

#### What I have learned working on this feature: 
More indepth knowledge of devise and adding more user roles.

#### Screenshots:
### Create article page:
![image](https://user-images.githubusercontent.com/45144030/54423989-35da6780-4712-11e9-98fc-d19402b0132a.png)

### Sign up/sign in:
![image](https://user-images.githubusercontent.com/45144030/54424217-b4cfa000-4712-11e9-91bb-d8262aa51841.png)

Joberg95 committed:
installed devise and generated model
migrated db

Joberg95 committed:
added create article page
migrated user model into database with role
added routes

perhegg committed:
create journalist can create article feature file, setup user factory and adds warder helpers

perhegg committed:
add create action to article controller and sets up routes, write feature tests

Joberg95 committed:
removed before_action :authenticate_user! from articles controller

Joberg95 committed:
removed unnecessary if statement in controller

perhegg committed:
remove extra feature tests

magnus-thor commented on app/views/articles/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+<%= notice %>
```
You are rendering the `notice` in application.html.haml and this will just duplicate the message

perhegg replied:
done

magnus-thor commented on config/routes.rb:
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
-    resources :articles, only: [:index]
+  devise_for :users
```
fix the indentation

perhegg replied:
done

magnus-thor commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,15 @@
+        Given we have 1 user with email "jocke@craft.se" and role "journalist"
+        And I am logged in as 'jocke@craft.se'
+        And I visit the journalist page
+        Then I fill in "Title" with "Vikings living amongst us"
```
This is a `when` step

perhegg replied:
done

magnus-thor commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,15 @@
+    I would like to be able to create an article
+
+    Scenario: Journalist can create article
+        Given we have 1 user with email "jocke@craft.se" and role "journalist"
```
Put this in the background and create a table for it.

perhegg replied:
done

magnus-thor commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,15 @@
+
+    Scenario: Journalist can create article
+        Given we have 1 user with email "jocke@craft.se" and role "journalist"
+        And I am logged in as 'jocke@craft.se'
```
Have this as a background step

perhegg replied:
done

magnus-thor commented on features/step_definitions/basic_steps.rb:
```diff
@@ -2,9 +2,30 @@
+end  
+  
+Given("we have (1 )user with email {string} and role {string}") do |email, role|
+    FactoryBot.create(:user, email: email, role: role)
```
You can ski the `FactoryBot` and just write
```
create(:user, email: email, role: role)
```

perhegg replied:
done

magnus-thor commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
add unit tests for the fields and the valid factory.

perhegg replied:
done

perhegg committed:
merge from upstream development

perhegg committed:
fix PR change requests

magnus-thor commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,19 @@
+Feature: Journalist can create articles
```
Add a sad path for a normal user should not be able to create a article.

Joberg95 committed:
added user to db in rails c to match tests and actual role
added before_action to check if journalist
changed migration from string to integer at role
added enums

perhegg committed:
add check to see if user is journalist,
create sad path for creating article, add sign_out button

Joberg95 commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,33 @@
+        And I fill in "Title" with "Vikings living amongst us"
+        And I fill in "Lead" with "Hurra"
+        And I click "Save Article"      
+        # Then I should not see "Article was successfully created."
```
We need some assistance on this step, it still shows "Article was successfully created" on the page even though we cant see it

magnus-thor replied:
Please use this step.

magnus-thor commented on app/models/article.rb:
```diff
@@ -1,5 +1,5 @@
 class Article < ApplicationRecord
     validates :title, presence: true
-    validates :content, presence: true
+    validates   :content, presence: true
```
extra space after validates

magnus-thor commented on app/views/layouts/application.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <body>
+    <%= notice || alert %>
+    <%= yield %>
+    <%= link_to "Sign out", destroy_user_session_path, :method => :delete %> 
```
old syntax `:method => :delete `

NiciP commented on app/controllers/articles_controller.rb:
```diff
@@ -1,9 +1,37 @@
-    
+
+    def new
+        
```
Remove this line?

Joberg95 committed:
removed spaces from model file, changed delete syntax to new delete syntax
removed feature test step

magnus-thor commented on app/controllers/articles_controller.rb:
```diff
@@ -1,9 +1,37 @@
         @article = Article.find(params[:id])
     end
+
+    def create
```
Rewrite this method like this and the sad path will work with not seeing the success message. Also are you sure you want to redirect to new_article_path after successfully creating article?
```
    def create
        article = Article.new(article_params)
       
        if article.save
            redirect_to new_article_path, notice: 'Article was successfully created.'
        else
            redirect_to new_article_path, alert: 'You have to fill out all the fields'
        end
    end
```

magnus-thor commented on app/views/layouts/application.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <body>
+    <%= notice || alert %>
+    <%= yield %>
+    <%= link_to "Sign out", destroy_user_session_path, method: :delete %>
```
Put a condition on this button to only show when user is logged in, and add a sign in button

Joberg95 committed:
added sign in/sign out condition to sign in button
refactored controller to fix sad path
sad path working

magnus-thor reviewed (approved):
Looks good
